File: Benchmarks_FSharp/Core/Entity/Yotsuba.fs
namespace Benchmarks.Core.Entity

open System

[<Flags>]
type YTBComponent =
    | Sprite = 1us
    | Transform = 2us
    | Animation = 4us
    | Rigibody = 8us
    | Input = 16us
    | Model3D = 32us
    | Button2D = 64us
    | Camera = 128us
    | Script = 256us
    | TileMap = 512us
    | Font = 1024us
    | Shader = 2048us
    | YTBUIElement = 4096us
    | YTBModel3D = 16384us

[<Struct>]
type Yotsuba =
    val mutable Id: int
    val mutable Name: string
    val mutable Components: int

    new(id: int) = { Id = id; Name = null; Components = 0 }

    member this.HasComponent(component: YTBComponent) =
        (this.Components &&& int component) <> 0

    member this.HasNotComponent(component: YTBComponent) =
        not (this.HasComponent(component))

    member this.AddComponent(component: YTBComponent) =
        this.Components <- this.Components ||| int component

    member this.RemoveComponent(component: YTBComponent) =
        this.Components <- this.Components &&& ~~~(int component)

File: Benchmarks/Core/Entity/Yotsuba.cs
using System;

namespace Benchmarks.Core.Entity
{
    /// <summary>
    /// Clase padre de todas las entidades. En este engine, a las entidades se les llaman "Yotsuba".
    /// </summary>
    public struct Yotsuba(int id)
    {
        public int Id { get; set; } = id;
        public string Name { get; set; }
        public int Components { get; set; } = 0;

        public bool HasComponent(YTBComponent component) => (Components & (int)component) != 0;
        public bool HasNotComponent(YTBComponent component) => !HasComponent(component);
        public void AddComponent(YTBComponent component) => Components |= (int)component;

        public void RemoveComponent(YTBComponent component)
        {
            Components &= ~(int)component;
        }
    }

    [Flags]
    public enum YTBComponent : ushort
    {
        Sprite = 1 << 0,
        Transform = 1 << 1,
        Animation = 1 << 2,
        Rigibody = 1 << 3,
        Input = 1 << 4,
        Model3D = 1 << 5,
        Button2D = 1 << 6,
        Camera = 1 << 7,
        Script = 1 << 8,
        TileMap = 1 << 9,
        Font = 1 << 10,
        Shader = 1 << 11,
        YTBUIElement = 1 << 12,
        YTBModel3D = 1 << 14,
    }
}
